### PR TITLE
UNINDEXED column constraint

### PR DESCRIPTION
**Pull request by fnc12**
This PR adds `UNINDEXED` column constraint which is used in FTS virtual table.
To improve: add static assertion in `table_t` to ensure that its columns do not contain `unindexed_t`.

### CODE REVIEW

File: dev/statement_serializer.h [hunk 1]
```diff
@@ -1140,12 +1150,17 @@ namespace sqlite_orm {
...
                 ss << streaming_identifier(column.name);
-                if(!context.skip_types_and_constraints) {
+                if(!context.skip_types_and_constraints_except_unindexed) {
                     ss << " " << type_printer<field_type_t<column_type>>().print();
                     ss << streaming_column_constraints(
                         call_as_template_base<column_constraints>(polyfill::identity{})(column),
                         column.is_not_null(),
                         context);
+                } else {
+                    using constraints_tuple = typename column_type::constraints_type;
```
Comment (fnc12) [inline dev/statement_serializer.h:1160]: @trueqbit probably this algorithm can be implemented way more beautiful. Do you have any idea?

File: tests/schema/virtual_table.cpp [hunk 2]
```diff
@@ -30,8 +30,6 @@ TEST_CASE("virtual table") {
         Post{"SQLite Tutorial", "Help you learn SQLite quickly and effectively"},
     };
 
-    //    storage.remove_all<Post>();
```
Comment (fnc12) [inline tests/schema/virtual_table.cpp:33]: removed old artifact

File: tests/statement_serializer_tests/schema/table.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#include <sqlite_orm/sqlite_orm.h>
+#include <catch2/catch_all.hpp>
+
+using namespace sqlite_orm;
+
+TEST_CASE("statement_serializer table_t") {
```
Comment (fnc12) [inline tests/statement_serializer_tests/schema/table.cpp:6]: this test case is not related to PRs topic but I wanted to add this last time in the next PR so here we are

File: tests/storage_tests.cpp [hunk 4]
```diff
@@ -268,3 +268,184 @@ TEST_CASE("non-unique DBOs") {
                                       make_column("time", &Record::setTime, &Record::time)));
     db.sync_schema();
 }
+
+TEST_CASE("insert") {
```
Comment (fnc12) [inline tests/storage_tests.cpp:272]: I moved all remaining tests from `tests2.cpp` here cause they are related to `storage_t`

File: dev/serializer_context.h [hunk 5]
```diff
@@ -8,7 +8,7 @@ namespace sqlite_orm {
             bool replace_bindable_with_question = false;
             bool skip_table_name = true;
             bool use_parentheses = true;
-            bool skip_types_and_constraints = false;
+            bool skip_types_and_constraints_except_unindexed = false;
```
Comment (trueqbit) [inline dev/serializer_context.h:11]: Can we rename this flag into something like `fts5_columns`? Also see below when checking it.

File: dev/statement_serializer.h (hunk 1 above)
Comment (trueqbit) [reply dev/statement_serializer.h:1160]: Yes, indeed ;)
```c++
        template<class G, class S, class... Op>
        struct statement_serializer<column_t<G, S, Op...>, void> {
            using statement_type = column_t<G, S, Op...>;

            template<class Ctx>
            std::string operator()(const statement_type& column, const Ctx& context) const {
                std::stringstream ss;
                ss << streaming_identifier(column.name);
                if(!context.fts5_columns) {
                    ss << " " << type_printer<field_type_t<column_field<G, S>>>().print();
                }
                ss << streaming_column_constraints(
                    call_as_template_base<column_constraints>(polyfill::identity{})(column),
                    column.is_not_null(),
                    context);
                return ss.str();
            }
        };
```

And the accompanying `streaming_column_constraints` in _serializing_util.h_:
```c++
        template<class... Op, class Ctx>
        std::ostream& operator<<(std::ostream& ss,
                                 std::tuple<const streaming<stream_as::column_constraints>&,
                                            const column_constraints<Op...>&,
                                            const bool&,
                                            Ctx> tpl) {
            const auto& column = std::get<1>(tpl);
            const bool& isNotNull = std::get<2>(tpl);
            auto& context = std::get<3>(tpl);

            using constraints_tuple = decltype(column.constraints);
            if(!context.fts5_columns) {
                iterate_tuple(column.constraints, [&ss, &context](auto& constraint) {
                    ss << ' ' << serialize(constraint, context);
                });
                constexpr bool hasExplicitNullableConstraint =
                    mpl::invoke_t<mpl::disjunction<check_if_has_type<null_t>, check_if_has_type<not_null_t>>,
                                  constraints_tuple>::value;
                if SQLITE_ORM_CONSTEXPR_IF(!hasExplicitNullableConstraint) {
                    if(isNotNull) {
                        ss << " NOT NULL";
                    } else {
                        ss << " NULL";
                    }
                }
            } else {
                constexpr bool hasUnindexedOption =
                    mpl::invoke_t<check_if_has_type<unindexed_t>, constraints_tuple>::value;
                if SQLITE_ORM_CONSTEXPR_IF(hasUnindexedOption) {
                    ss << " UNINDEXED";
                }
            }

            return ss;
        }
```

**Review by trueqbit — CHANGES_REQUESTED** [review]

File: dev/statement_serializer.h [hunk 6]
```diff
@@ -1045,6 +1045,16 @@ namespace sqlite_orm {
             }
         };
 
+        template<>
+        struct statement_serializer<unindexed_t, void> {
+            using statement_type = unindexed_t;
+
+            template<class Ctx>
+            std::string operator()(const statement_type& c, const Ctx& context) const {
```
[the remark this answers is not in the document]
Comment (fnc12) [reply dev/statement_serializer.h:1053]: done

File: tests/statement_serializer_tests/schema/table.cpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    std::string value;
+    std::string expected;
+    SECTION("simple") {
+        auto table = make_table("users", make_column("id", &User::id), make_column("name", &User::name));
+        using db_objects_t = internal::db_objects_tuple<decltype(table)>;
+        auto dbObjects = db_objects_t{table};
+        using context_t = internal::serializer_context<db_objects_t>;
+        context_t context{dbObjects};
+        value = internal::serialize(table, context);
+        expected = "CREATE TABLE \"users\" (\"id\" INTEGER NOT NULL, \"name\" TEXT NOT NULL)";
```
[the remark this answers is not in the document]
Comment (fnc12) [reply tests/statement_serializer_tests/schema/table.cpp:20]: done

File: tests/statement_serializer_tests/schema/table.cpp [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    }
+    SECTION("without_rowid") {
+        auto table =
+            make_table("users", make_column("id", &User::id), make_column("name", &User::name)).without_rowid();
+        using db_objects_t = internal::db_objects_tuple<decltype(table)>;
+        auto dbObjects = db_objects_t{table};
+        using context_t = internal::serializer_context<db_objects_t>;
+        context_t context{dbObjects};
+        value = internal::serialize(table, context);
+        expected = "CREATE TABLE \"users\" (\"id\" INTEGER NOT NULL, \"name\" TEXT NOT NULL) WITHOUT ROWID";
```
[the remark this answers is not in the document]
Comment (fnc12) [reply tests/statement_serializer_tests/schema/table.cpp:30]: done

File: tests/statement_serializer_tests/schema/using_fts5.cpp [hunk 9]
```diff
@@ -8,9 +8,19 @@ TEST_CASE("statement_serializer using_fts5") {
...
+    std::string expected;
     internal::db_objects_tuple<> storage;
     internal::serializer_context<internal::db_objects_tuple<>> context{storage};
-    auto node = using_fts5(make_column("title", &Post::title), make_column("body", &Post::body));
-    auto value = serialize(node, context);
-    REQUIRE(value == "USING FTS5(\"title\", \"body\")");
+    SECTION("simple") {
+        auto node = using_fts5(make_column("title", &Post::title), make_column("body", &Post::body));
+        value = serialize(node, context);
+        expected = "USING FTS5(\"title\", \"body\")";
```
[the remark this answers is not in the document]
Comment (fnc12) [reply tests/statement_serializer_tests/schema/using_fts5.cpp:18]: done

File: tests/statement_serializer_tests/schema/using_fts5.cpp [hunk 10]
```diff
@@ -8,9 +8,19 @@ TEST_CASE("statement_serializer using_fts5") {
...
+    }
+    SECTION("unindexed") {
+        auto node = using_fts5(make_column("title", &Post::title), make_column("body", &Post::body, unindexed()));
+        value = serialize(node, context);
+        expected = "USING FTS5(\"title\", \"body\" UNINDEXED)";
```
[the remark this answers is not in the document]
Comment (fnc12) [reply tests/statement_serializer_tests/schema/using_fts5.cpp:23]: done

File: tests/CMakeLists.txt [hunk 11]
```diff
@@ -41,7 +41,6 @@ add_executable(unit_tests
     xdestroy_handling.cpp
     sync_schema_tests.cpp
     tests.cpp
-    tests2.cpp
```
[the remark this answers is not in the document]
Comment (fnc12) [reply tests/CMakeLists.txt:44]: yeah. I can tell you why all ungrouped tests were divided up to several files. Cause Windows CI couldn't handle such a huge `.cpp` file to compile 🤦

File: dev/serializer_context.h (hunk 5 above)
Comment (fnc12) [reply dev/serializer_context.h]: done

**Review by trueqbit — APPROVED** [review]

File: dev/statement_serializer.h (hunk 1 above)
Comment (fnc12) [reply dev/statement_serializer.h]: thanks

**Pull request merged**